.test-infra/jenkins/Infrastructure.groovy

```diff
@@ -0,0 +1,111 @@
+ * limitations under the License.
+ */
+
+import CommonJobProperties as common
+import CommonTestProperties.SDK
+
+class Infrastructure {
+
+  static void prepareSDKHarness(def context, SDK sdk, String repositoryRoot, String dockerTag) {
+    context.steps {
+      String sdkName = sdk.name().toLowerCase()
+      String image = "${repositoryRoot}/${sdkName}"
+      String imageTag = "${image}:${dockerTag}"
+
+      shell("echo \"Building SDK harness for ${sdkName} SDK.\"")
+      gradle {
+        rootBuildScriptDir(common.checkoutDir)
+        common.setGradleSwitches(delegate)
+        tasks(":beam-sdks-${sdkName}-container:docker")
+        switches("-Pdocker-repository-root=${repositoryRoot}")
+        switches("-Pdocker-tag=${dockerTag}")
+      }
+      shell("echo \"Tagging Harness' image\"...")
+      shell("docker tag ${image} ${imageTag}")
+      shell("echo \"Pushing Harness' image\"...")
+      shell("docker push ${imageTag}")
+    }
+  }
+
+  static void prepareFlinkJobServer(def context, String repositoryRoot, String dockerTag) {
+    context.steps {
+      String image = "${repositoryRoot}/flink-job-server"
+      String imageTag = "${image}:${dockerTag}"
+
+      shell('echo "Building Flink job Server"')
+
+      gradle {
+        rootBuildScriptDir(common.checkoutDir)
+        common.setGradleSwitches(delegate)
+        tasks(":runners:flink:1.7:job-server-container:docker")
+        switches("-Pdocker-repository-root=${repositoryRoot}")
+        switches("-Pdocker-tag=${dockerTag}")
+      }
+
+      shell("echo \"Tagging Flink Job Server's image\"...")
+      shell("docker tag ${image} ${imageTag}")
+      shell("echo \"Pushing Flink Job Server's image\"...")
+      shell("docker push ${imageTag}")
+    }
+  }
+
+  static void setupFlinkCluster(def context, String clusterNamePrefix, Integer workerCount, String imagesToPull, String jobServerImage) {
+    String gcsBucket = 'gs://beam-flink-cluster'
+    String clusterName = getClusterName(clusterNamePrefix)
+    String artifactsDir="${gcsBucket}/${clusterName}"
+
+    context.steps {
+      environmentVariables {
+        env("GCLOUD_ZONE", "us-central1-a")
+        env("CLUSTER_NAME", clusterName)
+        env("GCS_BUCKET", gcsBucket)
+        env("FLINK_DOWNLOAD_URL", 'https://archive.apache.org/dist/flink/flink-1.7.0/flink-1.7.0-bin-hadoop28-scala_2.11.tgz')
+        env("FLINK_NUM_WORKERS", workerCount)
+        env("DETACHED_MODE", 'true')
+
+        if(imagesToPull) {
+          env("HARNESS_IMAGES_TO_PULL", imagesToPull)
+        }
+
+        if(jobServerImage) {
+          env("JOB_SERVER_IMAGE", jobServerImage)
+          env("ARTIFACTS_DIR", artifactsDir)
+        }
+      }
+
+      shell('echo Setting up flink cluster')
+      shell("cd ${common.makePathAbsolute('src/.test-infra/dataproc/')}; ./create_flink_cluster.sh")
+    }
+  }
+
+  static void teardownDataproc(def context, String jobName) {
+    context.publishers {
+      postBuildScripts {
```

Comment:
This requires a plugin - I created an issue for this: https://issues.apache.org/jira/browse/BEAM-7457 

The goal here is to cleanup the cluster regardless of the test result.